### DISCUSSION TITLE
Better handling of thumbnails

### DISCUSSION THREAD

**Discussion by Resumane** [directus/directus#22882] [Feature Requests]
### Summary

Let us set a few default thumbnail sizes (the only ones that a user could request from the server, the others should not be allowed) and automatically generate those thumbnails when the pictures are uploaded.

### Motivation

The problem this fixes is that generating pictures in thumbnail sizes is generally quite slow, much too slow for a normal user.

### Detailed Design

Create a thumbnail section in the Settings page, let's set several if needed, different sizes and formats.

### Requirements List

not needed

### Drawbacks

No drawbacks.

### Alternatives

No alternatives? The problem this fixes is that generating pictures in thumbnail sizes is generally quite slow, much too slow for a normal user.

### Adoption Strategy

It's a very minor change, that should be opt-in.

**Comment by rijkvanzanten** [comment]
> Let us set a few default thumbnail sizes (the only ones that a user could request from the server, the others should not be allowed)

This is already possible from settings today 👍🏻 

> The problem this fixes is that generating pictures in thumbnail sizes is generally quite slow, much too slow for a normal user.

This sounds a bit odd to me. Generating the asset only happens once the first time the asset is requested, and is then served from a static file every other time after that. While you could skip that generation step when it's already done on upload, I doubt it'll create much performance increase in the day to day usage of the API 🤔

**Reply by Resumane** [reply]
> This sounds a bit odd to me. Generating the asset only happens once the first time the asset is requested, and is then served from a static file every other time after that. While you could skip that generation step when it's already done on upload, I doubt it'll create much performance increase in the day to day usage of the API 🤔

The day-to-day usage of the API is perfectly fine once the thumbnails are already cached. The problem is that first time. If I have a carousel with 20 pictures, then it will take a VERY long time for the first user to go through it (they won't!).
The most basic solution is that I do it myself for each new item, but it's a massive waste of time. Or I would have to write a script to do that for me, but it's weird to have to do that for a basic feature.

I want to mention that, even though that's maybe not related, my first carousel didn't have lazy loading and would therefore request all the thumbnails at once, that would result in not getting any pictures back and CORS errors. Possibly it's because of the way my server is set, but having the thumbnails pre-generated would fix that problem regardless.

**Reply by Resumane** [reply]
I've just measured it at 12 seconds per picture. Imagine the time it takes to go through a 20 pictures carousel, unless that user is VERY interested there's no way they'll go through them all. Then another visitor comes, pushes a few more pictures before being discouraged and bounces off too.

![image](https://github.com/directus/directus/assets/21341870/558a17b8-1673-4cc6-b442-22b7e5107bf4)
Here's an illustration, keep in mind that these pictures were also resized and compressed before uploading, if they were 10mb each it probably would have been even worse.

**Comment by Resumane** [comment]
![image](https://github.com/directus/directus/assets/21341870/2faf6d6c-b408-4ecb-a752-205f6e4510ab)
Even adding some long timeouts in Nginx isn't enough, because Cloudflare limits us to 100s and there's no way to change that, so error 524. Hopefully, now you see why it's an issue to keep things as they are now.
